parquet/src/arrow/async_reader/mod.rs

```diff
@@ -613,8 +623,18 @@ where
                     .fetch(&mut self.input, predicate.projection(), selection)
                     .await?;
 
+                let mut cache_projection = predicate.projection().clone();
+                cache_projection.intersect(&projection);
                 let array_reader = ArrayReaderBuilder::new(&row_group)
-                    .build_array_reader(self.fields.as_deref(), predicate.projection())?;
+                    .build_array_reader_with_cache(
+                        self.fields.as_deref(),
+                        predicate.projection(),
+                        CacheOptions {
+                            projection_mask: &cache_projection,
+                            cache: row_group_cache.clone(),
+                            role: crate::arrow::array_reader::CacheRole::Producer,
+                        },
```

Comment:
structurally both here and below it might help to keep the creation ofthe `CacheOptions` into the cache itself so a reader of this code doesn't have to understand the innards of the cache


```suggestion
                        row_group_cache.producer_options(projection, predicate.proection())
```